File: src/utils/trajectory_viewer.cpp
#include "idocp/utils/trajectory_viewer.hpp"

#include <thread>
#include <chrono>
#include <cmath>

#include <boost/filesystem.hpp>

#include "pinocchio/gepetto/viewer.hpp"

#include "pinocchio/parsers/urdf.hpp"
#include "pinocchio/parsers/utils.hpp"
#include "pinocchio/algorithm/joint-configuration.hpp"
#include "gepetto/viewer/corba/client.hh"
#include "gepetto/viewer/corba/conversions.hh"


namespace idocp {

TrajectoryViewer::TrajectoryViewer(const std::string& path_to_urdf,
                                   const std::string& path_to_pkg)
  : force_radius_(0.015),
    force_length_(0.5),
    force_scale_(0.75),
    friction_cone_scale_(0.15) {
  pinocchio::urdf::buildModel(boost::filesystem::absolute(path_to_urdf).string(), 
                              model_);
  pinocchio::urdf::buildGeom(model_, 
                             boost::filesystem::absolute(path_to_urdf).string(), 
                             pinocchio::VISUAL, vmodel_, 
                             boost::filesystem::absolute(path_to_pkg).string());
  setForceProperties();
  setFrictionConeProperties();
  setCameraTransformDefault();
}


TrajectoryViewer::TrajectoryViewer(const std::string& path_to_urdf)
  : force_radius_(0.015),
    force_length_(0.5),
    force_scale_(0.75),
    friction_cone_scale_(0.15) {
  std::string path_to_pkg = boost::filesystem::absolute(path_to_urdf).string();
  for (int i=0; i<3; ++i) {
    const int file_name_length 
        = boost::filesystem::absolute(path_to_pkg).filename().string().size();
    for (int j=0; j<file_name_length; ++j) path_to_pkg.pop_back();
    path_to_pkg.pop_back();
  }
  pinocchio::urdf::buildModel(boost::filesystem::absolute(path_to_urdf).string(), 
                              model_);
  pinocchio::urdf::buildGeom(model_, 
                             boost::filesystem::absolute(path_to_urdf).string(), 
                             pinocchio::VISUAL, vmodel_, path_to_pkg);
  setForceProperties();
  setFrictionConeProperties();
  setCameraTransformDefault();
}


void TrajectoryViewer::setForceProperties() {
  force_radius_ = 0.015;
  force_length_ = 0.5;
  force_scale_ = 0.75;
  force_color_[0] = 1.0;
  force_color_[1] = 1.0;
  force_color_[2] = 0.0;
  force_color_[3] = 1.0;
}


void TrajectoryViewer::setFrictionConeProperties() {
  friction_cone_scale_ = 0.15;
  cone_color_[0] = 0.3;
  cone_color_[1] = 0.3;
  cone_color_[2] = 0.7;
  cone_color_[3] = 0.7;
  x_axis_ << 1, 0, 0;
}


void TrajectoryViewer::display(const std::vector<Eigen::VectorXd>& q_traj, 
                               const double sampling_period_in_sec) {
  pinocchio::gepetto::Viewer viewer(model_, &vmodel_, NULL);
  const bool success = viewer.initViewer("idocp::TrajectoryViewer");
  if(!success) {
    std::cout << "Failed in connecting to CORBA!!" << std::endl;
  }

  auto gui = gepetto::viewer::corba::gui();
  const auto window_id = gui->createWindow("idocp::TrajectoryViewer");
  gui->createScene("hpp-gui");
  gui->addSceneToWindow("hpp-gui", window_id);
  viewer.loadViewerModel("hpp-gui");
  for (int i=0; i<model_.nframes; ++i) {
    viewer.addFrame(i);
  }

  if (!gui->nodeExists("hpp-gui/floor")) {
    gui->addFloor("hpp-gui/floor");
  }
  gui->setColor("hpp-gui/floor", gepetto::viewer::corba::grey);
  gui->setBackgroundColor1("idocp::TrajectoryViewer", gepetto::viewer::corba::white);
  gui->setBackgroundColor2("idocp::TrajectoryViewer", gepetto::viewer::corba::white);
  gui->setLightingMode("hpp-gui/floor", "OFF");

  setCameraTransform();

  for (const auto& q : q_traj) {
    viewer.display(q);
    std::this_thread::sleep_for(std::chrono::duration<double>(sampling_period_in_sec));
  }
}


void TrajectoryViewer::display(Robot& robot, 
                               const std::vector<Eigen::VectorXd>& q_traj, 
                               const std::vector<Eigen::VectorXd>& f_traj, 
                               const double sampling_period_in_sec,
                               const double mu) {
  pinocchio::gepetto::Viewer viewer(model_, &vmodel_, NULL);
  const bool success = viewer.initViewer("idocp::TrajectoryViewer");
  if(!success) {
    std::cout << "Failed in connecting to CORBA!!" << std::endl;
  }

  auto gui = gepetto::viewer::corba::gui();
  const auto window_id = gui->createWindow("idocp::TrajectoryViewer");
  gui->createScene("hpp-gui");
  gui->addSceneToWindow("hpp-gui", window_id);
  viewer.loadViewerModel("hpp-gui");
  for (int i=0; i<model_.nframes; ++i) {
    viewer.addFrame(i);
  }

  if (!gui->nodeExists("hpp-gui/floor")) {
    gui->addFloor("hpp-gui/floor");
  }
  gui->setColor("hpp-gui/floor", gepetto::viewer::corba::grey);
  gui->setBackgroundColor1("idocp::TrajectoryViewer", gepetto::viewer::corba::white);
  gui->setBackgroundColor2("idocp::TrajectoryViewer", gepetto::viewer::corba::white);
  gui->setLightingMode("hpp-gui/floor", "OFF");

  if (mu > 0) {
    gui->createGroup("hpp-gui/friction_cones");
    Eigen::Vector3d cone;
    cone << mu, mu, 1.0;
    for (int j=0; j<robot.maxPointContacts(); ++j) {
        gui->createGroup(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)).c_str());
        gui->addCurve(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/vertex").c_str(), 
                      gepetto::viewer::corba::positionSeq({ {0., 0., 0., }, 
                                                            { cone[0],  cone[1],  cone[2]}, 
                                                            {-cone[0],  cone[1],  cone[2]}, 
                                                            {-cone[0], -cone[1],  cone[2]}, 
                                                            { cone[0], -cone[1],  cone[2]}, 
                                                            { cone[0],  cone[1],  cone[2]}, }), cone_color_);
        gui->setCurveMode(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/vertex").c_str(), "TRIANGLE_FAN");
        gepetto::corbaserver::Position p0, p;
        p0[0] = 0.; p0[1] = 0.; p0[2] = 0.;
        p[0] =  cone[0]; p[1] =  cone[1]; p[2] = cone[2];
        gui->addLine(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line1").c_str(), p0, p, cone_color_);
        p[0] = -cone[0]; p[1] =  cone[1]; p[2] = cone[2];
        gui->addLine(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line2").c_str(), p0, p, cone_color_);
        p[0] = -cone[0]; p[1] = -cone[1]; p[2] = cone[2];
        gui->addLine(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line3").c_str(), p0, p, cone_color_);
        p[0] =  cone[0]; p[1] = -cone[1]; p[2] = cone[2];
        gui->addLine(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line4").c_str(), p0, p, cone_color_);
        gui->setFloatProperty(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)).c_str(), "Alpha", 0);
        gepetto::corbaserver::Position cone_scale;
        cone_scale[0] = friction_cone_scale_; cone_scale[1] = friction_cone_scale_; cone_scale[2] = friction_cone_scale_;
        gui->setScale(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/vertex").c_str(), cone_scale);
        gui->setScale(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line1").c_str(), cone_scale);
        gui->setScale(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line2").c_str(), cone_scale);
        gui->setScale(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line3").c_str(), cone_scale);
        gui->setScale(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line4").c_str(), cone_scale);
        gui->setFloatProperty(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/vertex").c_str(), "Alpha", 0);
        gui->setFloatProperty(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line1").c_str(), "Alpha", 0.2);
        gui->setFloatProperty(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line2").c_str(), "Alpha", 0.2);
        gui->setFloatProperty(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line3").c_str(), "Alpha", 0.2);
        gui->setFloatProperty(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)+"/line4").c_str(), "Alpha", 0.2);
    }
  }

  gui->createGroup("hpp-gui/contact_forces");
  for (int j=0; j<robot.maxPointContacts(); ++j) {
    gui->addArrow(("hpp-gui/contact_forces/contact_force_"+std::to_string(j)).c_str(), force_radius_, force_length_, force_color_);
    gui->setFloatProperty(("hpp-gui/contact_forces/contact_force_"+std::to_string(j)).c_str(), "Alpha", 1.0);
    gui->setVisibility(("hpp-gui/contact_forces/contact_force_"+std::to_string(j)).c_str(), "ALWAYS_ON_TOP");
  }

  setCameraTransform();

  for (int i=0; i<f_traj.size(); ++i) {
    robot.updateFrameKinematics(q_traj[i]);
    for (int j=0; j<robot.maxPointContacts(); ++j) {
      const Eigen::Vector3d& f = f_traj[i].template segment<3>(3*j);
      gepetto::corbaserver::Position f_scale;
      f_scale[0] = force_scale_ * std::sqrt(f.norm() / robot.totalWeight());
      f_scale[1] = 1.;
      f_scale[2] = 1.;
      gui->setVector3Property(("hpp-gui/contact_forces/contact_force_"+std::to_string(j)).c_str(), "Scale", f_scale);
      const Eigen::Matrix3d R = rotationMatrix(f);
      const pinocchio::SE3::Quaternion quat = pinocchio::SE3::Quaternion(R);
      gepetto::corbaserver::Transform pose;
      const int contact_frame = robot.contactFrames()[j];
      pose[0] = robot.framePosition(contact_frame)[0];
      pose[1] = robot.framePosition(contact_frame)[1];
      pose[2] = robot.framePosition(contact_frame)[2];
      pose[3] = quat.x();
      pose[4] = quat.y();
      pose[5] = quat.z();
      pose[6] = quat.w();
      gui->applyConfiguration(("hpp-gui/contact_forces/contact_force_"+std::to_string(j)).c_str(), pose);
      if (mu > 0) {
        pose[0] = robot.framePosition(contact_frame)[0];
        pose[1] = robot.framePosition(contact_frame)[1];
        pose[2] = robot.framePosition(contact_frame)[2];
        pose[3] = 0.;
        pose[4] = 0.;
        pose[5] = 0.;
        pose[6] = 1.;
        gui->applyConfiguration(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)).c_str(), pose);
      }
      if (f.norm() > 0) {
        if (mu > 0) {
          gui->setVisibility(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)).c_str(), "ON");
        }
        gui->setVisibility(("hpp-gui/contact_forces/contact_force_"+std::to_string(j)).c_str(), "ON");
      }
      else {
        if (mu > 0) {
          gui->setVisibility(("hpp-gui/friction_cones/friction_cone_"+std::to_string(j)).c_str(), "OFF");
        }
        gui->setVisibility(("hpp-gui/contact_forces/contact_force_"+std::to_string(j)).c_str(), "OFF");
      }
      gui->refresh();
    }
    viewer.display(q_traj[i]);
    std::this_thread::sleep_for(std::chrono::duration<double>(sampling_period_in_sec));
  }
}


void TrajectoryViewer::setCameraTransform(const Eigen::Vector3d& camera_pos,
                                          const Eigen::Vector4d& camera_quat) {
  camera_pos_ = camera_pos;
  camera_quat_ = camera_quat;
}


void TrajectoryViewer::setCameraTransformDefault() {
  camera_pos_ << 2.2, -3.5, 1.13;
  camera_quat_ << 0.60612, 0.166663, 0.19261, 0.753487;
}


void TrajectoryViewer::printCurrentCameraTransform() const {
  auto gui = gepetto::viewer::corba::gui();
  const auto window_id = gui->getWindowID("idocp::TrajectoryViewer");
  const auto camera = gui->getCameraTransform(window_id);
  std::cout << "Current camera transform is: [";
  for (int i=0; i<6; ++i) {
    std::cout << camera[i] << ", ";
  }
  std::cout << camera[6] << "]" << std::endl;
}


void TrajectoryViewer::setCameraTransform() {
  auto gui = gepetto::viewer::corba::gui();
  const auto window_id = gui->getWindowID("idocp::TrajectoryViewer");
  std::vector<float> pose(7);
  pose[0] = camera_pos_.coeff(0);
  pose[1] = camera_pos_.coeff(1);
  pose[2] = camera_pos_.coeff(2);
  pose[3] = camera_quat_.normalized().coeff(0);
  pose[4] = camera_quat_.normalized().coeff(1);
  pose[5] = camera_quat_.normalized().coeff(2);
  pose[6] = camera_quat_.normalized().coeff(3);
  gui->setCameraTransform(window_id, pose.data());
}

} // namespace idocp

File: bindings/python/idocp/utils/trajectory_viewer.py
import idocp
import pinocchio
from pinocchio.robot_wrapper import RobotWrapper
from pinocchio.visualize import GepettoVisualizer, MeshcatVisualizer
from os.path import abspath, dirname, join
import numpy as np
import math
import time
import meshcat.transformations 

class TrajectoryViewer:
    def __init__(self, path_to_urdf, path_to_pkg=None, 
                 base_joint_type=idocp.BaseJointType.FixedBase, 
                 viewer_type='gepetto'):
        self.path_to_urdf = abspath(path_to_urdf)
        if path_to_pkg is None:
            path_to_pkg = join(dirname(self.path_to_urdf), '../..')
        if base_joint_type == idocp.BaseJointType.FloatingBase:
            self.robot= RobotWrapper.BuildFromURDF(self.path_to_urdf, path_to_pkg,
                                                   pinocchio.JointModelFreeFlyer())
        else:
            self.robot= RobotWrapper.BuildFromURDF(self.path_to_urdf, path_to_pkg)
        self.path_to_urdf = path_to_urdf
        self.base_joint_type = base_joint_type

        self.play_speed = 1.0

        self.force_radius = 0.015
        self.force_length = 0.5
        self.force_scale = 0.75
        self.force_color = [1.0, 1.0, 0.0, 1.0]

        self.display_contact = False
        self.contact_frames = []
        self.mu = 0.
        self.cone_scale = [0.15, 0.15, 0.15]
        self.cone_color = [0.3, 0.3, 0.7, 0.7]
        self.x_axis = np.array([1.0, 0.0, 0.0])

        self.viewer_type = viewer_type
        if viewer_type == 'gepetto':
            self.viewer = GepettoVisualizer(self.robot.model, 
                                            self.robot.collision_model,
                                            self.robot.visual_model)
            self.window_name = 'idocp.TrajectoryViewer'
            self.camera_pos = [2.2, -3.5, 1.13] 
            self.camera_angle = [0.60612, 0.166663, 0.19261, 0.753487] 
        elif viewer_type == 'meshcat':
            self.viewer = MeshcatVisualizer(self.robot.model, 
                                            self.robot.collision_model, 
                                            self.robot.visual_model)
            self.camera_tf = meshcat.transformations.translation_matrix([0.8, -2.5, -0.2]) 
            self.zoom = 3.0
        else:
            print('Please choose viewer_type from "gepetto" or "meshcat"!')
            return NotImplementedError()


    def set_contact_info(self, contact_frames, mu):
        self.display_contact = True
        self.contact_frames = contact_frames
        self.mu = mu


    def set_camera_transform_gepetto(self, camera_pos=None, camera_angle=None):
        if camera_pos is not None:
            self.camera_pos = camera_pos
        if camera_angle is not None:
            self.camera_angle = camera_angle

    def set_camera_transform_meshcat(self, camera_tf_vec=None, zoom=None):
        if camera_tf_vec is not None:
            self.camera_tf = meshcat.transformations.translation_matrix(camera_tf_vec) 
        if zoom is not None:
            self.zoom = zoom


    def display(self, dt, q_traj, f_traj=None):
        if self.viewer_type == 'gepetto':
            self.display_gepetto(dt, q_traj, f_traj)
        elif self.viewer_type == 'meshcat':
            self.display_meshcat(dt, q_traj)


    def display_gepetto(self, dt, q_traj, f_traj):
        self.robot.setVisualizer(self.viewer)
        self.robot.initViewer(windowName=self.window_name, loadModel=False)
        self.robot.loadViewerModel(rootNodeName=self.window_name)
        gui = self.robot.viewer.gui
        # init the window
        window_id = gui.getWindowID(self.window_name)
        gui.setBackgroundColor1(window_id, [1., 1., 1., 1.])
        gui.setBackgroundColor2(window_id, [1., 1., 1., 1.])
        # init the floor
        floor_name = 'world/floor'
        gui.addFloor(floor_name)
        gui.setColor(floor_name, [0.7, 0.7, 0.7, 1.0])
        gui.setLightingMode(floor_name, 'OFF')
        # init contact forces and friction cones
        if f_traj is not None and self.display_contact:
            # create cones
            self.robot.viewer.gui.createGroup('world/friction_cones')
            cone = [self.mu, self.mu, 1.0]
            for i in range(len(self.contact_frames)):
                gui.createGroup('world/friction_cones/friction_cone_'+str(i))
                gui.addCurve('world/friction_cones/friction_cone_'+str(i)+'/vertex', 
                             [[0., 0., 0.], 
                             [ cone[0],  cone[1],  cone[2]],
                             [-cone[0],  cone[1],  cone[2]],
                             [-cone[0], -cone[1],  cone[2]],
                             [ cone[0], -cone[1],  cone[2]],
                             [ cone[0],  cone[1],  cone[2]]], self.cone_color)
                gui.setCurveMode('world/friction_cones/friction_cone_'+str(i)+'/vertex', 'TRIANGLE_FAN')
                gui.addLine('world/friction_cones/friction_cone_'+str(i)+'/line1', 
                            [0., 0., 0.], [ cone[0],  cone[1],  cone[2]], self.cone_color)
                gui.addLine('world/friction_cones/friction_cone_'+str(i)+'/line2', 
                            [0., 0., 0.], [-cone[0],  cone[1],  cone[2]], self.cone_color)
                gui.addLine('world/friction_cones/friction_cone_'+str(i)+'/line3', 
                            [0., 0., 0.], [-cone[0], -cone[1],  cone[2]], self.cone_color)
                gui.addLine('world/friction_cones/friction_cone_'+str(i)+'/line4', 
                            [0., 0., 0.], [ cone[0], -cone[1],  cone[2]], self.cone_color)
                gui.setScale('world/friction_cones/friction_cone_'+str(i)+'/vertex', self.cone_scale)
                gui.setScale('world/friction_cones/friction_cone_'+str(i)+'/line1', self.cone_scale)
                gui.setScale('world/friction_cones/friction_cone_'+str(i)+'/line2', self.cone_scale)
                gui.setScale('world/friction_cones/friction_cone_'+str(i)+'/line3', self.cone_scale)
                gui.setScale('world/friction_cones/friction_cone_'+str(i)+'/line4', self.cone_scale)
                gui.setFloatProperty('world/friction_cones/friction_cone_'+str(i)+'/vertex', 'Alpha', 0.)
                gui.setFloatProperty('world/friction_cones/friction_cone_'+str(i)+'/line1', 'Alpha', 0.2)
                gui.setFloatProperty('world/friction_cones/friction_cone_'+str(i)+'/line2', 'Alpha', 0.2)
                gui.setFloatProperty('world/friction_cones/friction_cone_'+str(i)+'/line3', 'Alpha', 0.2)
                gui.setFloatProperty('world/friction_cones/friction_cone_'+str(i)+'/line4', 'Alpha', 0.2)
            # create forces
            gui.createGroup('world/contact_forces')
            for i in range(len(self.contact_frames)):
                gui.addArrow('world/contact_forces/contact_force_'+str(i),
                             self.force_radius, self.force_length, self.force_color)
                gui.setFloatProperty('world/contact_forces/contact_force_'+str(i), 'Alpha', 1.0)
                gui.setVisibility('world/contact_forces/contact_force_'+str(i), 'ALWAYS_ON_TOP')
        # set camera and play speed
        sleep_time = dt / self.play_speed
        camera = self.camera_pos
        camera.extend(self.camera_angle)
        gui.setCameraTransform(self.robot.viz.windowID, camera)
        # display
        if f_traj is not None:
            robot = idocp.Robot(self.path_to_urdf, self.base_joint_type, 
                                self.contact_frames)
            for q, f in zip(q_traj, f_traj):
                robot.forward_kinematics(q)
                for i in range(len(self.contact_frames)):
                    fi = f[3*i:3*(i+1)]
                    f_scale = [math.sqrt(self.force_scale*np.linalg.norm(fi)/robot.total_weight()), 1.0, 1.0]
                    gui.setVector3Property('world/contact_forces/contact_force_'+str(i), "Scale", f_scale)
                    fpos = robot.frame_position(self.contact_frames[i])
                    quat = pinocchio.Quaternion(self.x_axis, fi).normalized()
                    pose = np.concatenate((fpos, np.array([quat.x, quat.y, quat.z, quat.w])), axis=None).tolist()
                    gui.applyConfiguration('world/contact_forces/contact_force_'+str(i), pose)
                    if self.mu > 0.0:
                        pose = np.concatenate((fpos, np.array([0., 0., 0., 1.])), axis=None).tolist()
                        gui.applyConfiguration('world/friction_cones/friction_cone_'+str(i), pose)
                    if np.linalg.norm(fi) > 0.0 and self.mu > 0.0:
                        gui.setVisibility('world/friction_cones/friction_cone_'+str(i), 'ON')
                        gui.setVisibility('world/contact_forces/contact_force_'+str(i), 'ON')
                    else:
                        gui.setVisibility('world/friction_cones/friction_cone_'+str(i), 'OFF')
                        gui.setVisibility('world/contact_forces/contact_force_'+str(i), 'OFF')
                    gui.refresh()
                self.robot.display(q)
                time.sleep(sleep_time)
        else:
            for q in q_traj:
                self.robot.display(q)
                time.sleep(sleep_time)


    def display_meshcat(self, dt, q_traj, open=True):
        self.robot.setVisualizer(self.viewer)
        self.robot.initViewer(open=open)
        self.robot.loadViewerModel(rootNodeName='idocp.TrajectoryViewer')
        self.viewer.viewer["/Cameras/default"].set_transform(self.camera_tf)
        self.viewer.viewer["/Cameras/default/rotated/<object>"].set_property("zoom", self.zoom)
        # self.viewer.viewer["/Cameras/default"].set_transform(tf.translation_matrix([0.3, -2.5, -0.2]))
        # self.viewer.viewer["/Cameras/default/rotated/<object>"].set_property("zoom", 5.0)

        sleep_time = dt / self.play_speed
        for q in q_traj:
            self.robot.display(q)
            time.sleep(sleep_time)